lib/cn/cn_tree.c

```diff
@@ -1308,13 +1316,15 @@ cn_comp_update_kvcompact(struct cn_compaction_work *work, struct kvset *new_kvse
         return;
 
     INIT_LIST_HEAD(&retired_kvsets);
-    assert(work->cw_dgen_hi_min == kvset_get_workid(work->cw_mark->le_kvset));
 
     rmlock_wlock(&tree->ct_lock);
     {
-        assert(!list_empty(&work->cw_node->tn_kvset_list));
         le = work->cw_mark;
         for (i = 0; i < work->cw_kvset_cnt; i++) {
+            assert(&le->le_link != &work->cw_node->tn_kvset_list);
```

Comment:
Prior to your patch, we didn't require resetting the workid for these kvsets which are marked for deletion.

I presume you are resetting work now as the work struct can get destroyed before the kvset gets destroyed?

Is it done here

Reply:
Yeah, my intent was to prevent someone from dereferencing the work ptr after the kvset has been removed from the node, but is still alive due to refcnt > 0.  But I couldn't do that everwhere (e.g., in split), so I think I'll just revert these and document that you can only deref it if the kvset belongs to a node.
That said, I still want to make the check here :-)

Reply:
I've removed all the unnecessary "set_work()" calls in my upcoming commit, but have left the new asserts to check that the work Id is as expected.

Reply:
Nobody likes invalid ptrs (work) lingering in valid structs (kvsets).
What's the motivation for work ptr to replace work id?

Reply:
Debugging.  When I hit the problem the dgen didn't help me, and actually confused me greatly, because it was the dgen I wanted I was intending to set it to, so I ended up going in circles trying to figure out why.
Using the work compaction ptr is unique while the kvset is in the node, which better ties together the kvsets from two nodes (e.g. join), and if you're in the debugger you can now easily find the job that is working on that kvset.
A big problem with all the previous asserts that checked workid is that they only tested against zero; testing against the expected dgen would have been a much stronger test, but wouldn't have helped in this case since the dgens were the identical.
I would prefer to clear the ptr when we remove a kvset from the node, but split didn't make that easy because it just move the list rather than iterating over it.  So what's a mother to do?

Reply:
- which better ties together the kvsets from two nodes (e.g. join)
Good point.

So is `kvset->ks_work` derefenced anywhere?  If not, maybe make it a `void *` or a `uint64_t` so that is is opaque.

Reply:
Just one place, in kvdb_rest to gather current compaction status (while holding the tree lock).
ks_work is opaque, kvset doesn't know anything about it, kvset just provides the "get/sec work" interface for csched/compaction to mark busy kvsets.  Previously it was a uint64_t, now it's a const void *.

Reply:
> now it's a const void *.

sweet.

Reply:
If we had smart ptrs like C++ it'd be easy!